src/TufValidatedComposerRepository.php

```diff
@@ -88,9 +89,52 @@ protected function configurePackageTransportOptions(PackageInterface $package)
             'repository' => $config['url'],
             'target' => $package->getName() . '/' . $package->getVersion(),
         ];
+        if ($this->updater) {
+            try {
+                $options['max_file_size'] = $this->updater->getLength($options['tuf']['target']);
+            } catch (NotFoundException $e) {
+                $options['max_file_size'] = Updater::MAXIMUM_DOWNLOAD_BYTES;
+            }
+        }
         $package->setTransportOptions($options);
     }
 
+    /**
+     * Extracts a TUF target path from a full URL.
+     *
+     * @param string $url
+     *   A URL.
+     *
+     * @return string
+     *   A TUF target path derived from the URL.
+     */
+    private function getTargetFromUrl(string $url): string
+    {
+        $config = $this->getRepoConfig();
+        $target = str_replace($config['url'], null, $url);
+        return ltrim($target, '/');
+    }
+
+    /**
+     * Reacts before metadata is downloaded.
+     *
+     * @param PreFileDownloadEvent $event
+     *   The event object.
+     */
+    public function prepareMetadata(PreFileDownloadEvent $event): void
+    {
+        if ($this->updater) {
+            $target = $this->getTargetFromUrl($event->getProcessedUrl());
+            $options = $event->getTransportOptions();
+            try {
+                $options['max_file_size'] = $this->updater->getLength($target);
+            } catch (NotFoundException $e) {
+                $options['max_file_size'] = Updater::MAXIMUM_DOWNLOAD_BYTES;
+            }
```

Comment:
In what case would the `NotFoundException` not a reason we should stop the update process. We should always find a target correct?

Reply:
Yup, I thought about this just before you posted your review and I ended up changing it for packages, but not for metadata, due to the way ComposerRepository works. I added a long comment explaining the reasoning.